openmc/universe.py

```diff
@@ -396,6 +401,35 @@ def plot(self, origin=(0., 0., 0.), width=(1., 1.), pixels=(200, 200),
                 height = pixels[0]*px/(params.top - params.bottom)
                 fig.set_size_inches(width, height)
 
+            # add legend showing which colors represent which material
+            # or cell if that was requested
+            if legend:
+                if plot.colors is None:
+                    raise Exception("Must set plot color dictionary if you "
+                                    "would like to have an automatic legend.")
+
+                if color_by == "cell":
+                    expected_key_type = openmc.Cell
+                else:
+                    expected_key_type = openmc.Material
+
+                patches = []
+                for key, color in plot.colors.items():
+
+                    if isinstance(key, int):
+                        raise Exception(
+                            "Cannot use IDs in colors dict for auto legend.")
+                    elif not isinstance(key, expected_key_type):
+                        raise Exception(
+                            "Color dict key type does not match color_by")
+
+                    # this works whether we're doing cells or materials
+                    label = key.name if key.name != '' else key.id
+                    key_patch = mpatches.Patch(color=color, label=label)
+                    patches.append(key_patch)
+
+                axes.legend(handles=patches, bbox_to_anchor=(1.05, 1), loc=2, borderaxespad=0.0)
```

Comment:
Perhaps we ought to allow a `legend_kwargs` argument that allows the user to customize it as they wish

Reply:
epic idea. will give it a try.